### DISCUSSION TITLE
YAML include/import

### DISCUSSION THREAD

**Discussion by Chris-Donovan** [hmatuschek/qdmr#409] [Ideas]
I've just started playing around with qdmr and I love the ability to edit/store radio configurations in YAML format. I like the ability to easily diff and potentially version control my radio configs.

I have multiple radios and I would like to avoid duplicating effort across multiple config files by defining the same channels and zones in each configuration. Have you considered importing additional YAML files so you can define the channels and zones outside of the main config file?

This would allow the user to reuse a group of channels and zones across multiple radio configs and also allow for groups of channels and zones to be easily added or removed from a codeplug as circumstances change.

If I were to work on prototyping something, would it be more appropriate to do this within the qdmr code base or as an external application that parses the YAML files and generates a single qdmr compatible YAML file?

**Comment by allesand** [comment]
@Chris-Donovan : actually, you can (to the extend of the radios capabilities) use the *same* config file on all your radios - that's what qdmr was intended for. This works as long as for examples number of channels in a zone or numbers of entries in an RX group is supported on all radios. How different are your radios in that area (32 max entries in an RX group vs. 64 max etc.)?

@hmatuschek : since the reference IDs change on insert/delete, it would be hard to do import/export? Would an "enable"/"disable" flag for Channels or Zones or Groups be an option?

**Reply by Chris-Donovan** [reply]
@allesand: In my case, I have an Anytone D878UV and a D878UVII. Some use cases I would foresee:

- Configuring APRS receive on only one radio (not supported on both)
- Configuring a different APRS SSIDs for the two radios
- Configuring different DMR IDs for the other HAMs in the house
- Quickly adding/removing channels/groups for a special event (e.g. event comms, travel, etc.)

In most cases I would envision the diff between the two radios would be on the order of 10s of lines of YAML while the adding and subtracting of channels might have be much larger, it would be nice to be able to remove them from the config while still saving them for next year.

A lot of the components (channels, etc.) in the config I pull from my radio have anytone specific options. How does qdmr handle radio specific settings for multiple radios? Do the irrelevant ones get ignored? (e.g. ignoring the anytone section on a non-anytone radio)

**Comment by allesand** [comment]
@Chris-Donovan : well, almost:

> In my case, I have an Anytone D878UV and a D878UVII. Some use cases I would foresee:

Same here :-)

> * Configuring APRS receive on only one radio (not supported on both)

If you want to share configs between the two, you can only *write* with qdmr (and not read from the radios). Since qdmr knows that an UV does not support analog APRS receive, it will skip that part of the config.That's exactly how I configure my radios. Same config for the UV and the UVII, I have to admit that I only send digital APRS. But for example writing the user DB just works because qdmr knows how many entries fit in the radios.

> * Configuring a different APRS SSIDs for the two radios

That is a manual process, I guess, indeed.

> * Configuring different DMR IDs for the other HAMs in the house

If you enable "commercial features" with the gear icon, you can chose the default Radio ID on the corresponding tab - if all channels are set to [Default] DMR ID, that should work.

> * Quickly adding/removing channels/groups for a special event (e.g. event comms, travel, etc.)

That is a problem, yes.. Zones and channels just keep accumulating :-) The "Zone hide" feature in the Anytone helps somewhat.

> A lot of the components (channels, etc.) in the config I pull from my radio have anytone specific options. How does qdmr handle radio specific settings for multiple radios? Do the irrelevant ones get ignored? (e.g. ignoring the anytone section on a non-anytone radio)

@hmatuschek : from what I understand, qdmr always checks which config values fit in the target radio? Worked in all my tests with different radios so far. What does *not* work when for example the radio supports only 32 entries in an RX group (RD-5R) - the Anytone config does not work in that case.

Depending on what qdmr version you are using right now, you might be hit by #375 - it is fixed in master and will be in 0.12..

**Comment by hmatuschek** [comment]
## Concerning Limits
qdmr never exceeds any limits imposed by any radio. This is ensured during the encoding of the codeplugs for every device separately. That is, if there is a 32 entry limit for group lists on the RD-5R only 32 will be written. However, you are not directly able to control which 32 are written (you are to some degree). Before that happens, however, there are several lines of defense:

1.  If some cosmetics do not fit, a notification is issued. For example, names are usually limited to 16 or 32 chars. Some radios limit some names to only 8 chars. qdmr truncates the name and issues a notification.
2. If some limits are exceeded that may impact the functionality of the codeplug, a warning is issued. This is for example, if a group list exceeds the number of elements for the target device. 
3. If some limits are exceeded, that may danger the device (e.g. frequency limits) an error is issued and the codeplug upload is prevented. Unless these errors are overridden by an settings option. 

## Concerning YAML merges
Thought about that too. Maybe not automatically but some semi-automatic means to import/merge codeplugs in qdmr. If, however, you want to do everything automatically, consider just writing some simple python script merging YAML files as you like it. You may then even use the command line tool to upload the generated YAML files, bypassing the GUI entirely.

**Comment by hmatuschek** [comment]
Import/merging of code plugs is now implemented in the devel branch and will be part of the next release.
